File: backend/app/schemas/lead.py
import uuid
from datetime import datetime, timezone
from typing import Optional, List

from pydantic import BaseModel, Field, computed_field
from app.schemas.assessment import AssessmentOut


class LeadIngest(BaseModel):
    copper_id: Optional[str] = None
    company_name: str
    website: Optional[str] = None
    description: Optional[str] = None
    stage: Optional[str] = None
    region: Optional[str] = None
    founder_names: Optional[List[str]] = None
    linkedin_urls: Optional[List[str]] = None
    raw_copper_data: Optional[dict] = None


class LeadUpdate(BaseModel):
    status: Optional[str] = None
    description: Optional[str] = None
    company_linkedin_url: Optional[str] = None


class LeadOut(BaseModel):
    id: uuid.UUID
    copper_id: Optional[str]
    owner_email: Optional[str]
    company_name: str
    website: Optional[str]
    description: Optional[str]
    stage: Optional[str]
    region: Optional[str]
    founder_names: Optional[List[str]]
    linkedin_urls: Optional[List[str]]
    company_linkedin_url: Optional[str]
    pitch_deck_filename: Optional[str] = None
    pitch_deck_ingested_at: Optional[datetime] = None
    pitch_deck_drive_id: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: datetime
    # Only used to derive applied_at below — never serialized to the API.
    raw_copper_data: Optional[dict] = Field(default=None, exclude=True)

    model_config = {"from_attributes": True}

    @computed_field  # type: ignore[misc]
    @property
    def applied_at(self) -> Optional[datetime]:
        """The lead's true application date from Copper, falling back to our
        import timestamp (created_at) when Copper's date isn't available —
        raw_copper_data["date_created"] is epoch seconds from Copper's API."""
        raw = self.raw_copper_data
        if isinstance(raw, dict):
            date_created = raw.get("date_created")
            if isinstance(date_created, (int, float)) and not isinstance(date_created, bool):
                try:
                    return datetime.fromtimestamp(date_created, tz=timezone.utc)
                except (ValueError, OSError, OverflowError):
                    pass
        return self.created_at


class LeadWithAssessment(LeadOut):
    assessment: Optional[AssessmentOut] = None

    model_config = {"from_attributes": True}


class PaginatedLeads(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[LeadWithAssessment]

File: frontend/src/types/lead.ts
export interface Lead {
  id: string;
  copper_id: string | null;
  company_name: string;
  website: string | null;
  description: string | null;
  stage: string | null;
  region: string | null;
  founder_names: string[] | null;
  linkedin_urls: string[] | null;
  company_linkedin_url: string | null;
  pitch_deck_filename: string | null;
  pitch_deck_ingested_at: string | null;
  pitch_deck_drive_id: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  applied_at: string | null;
  assessment?: Assessment;
}

export interface Assessment {
  id: string;
  lead_id: string;
  bucket: "YES" | "MAYBE" | "REJECT";
  confidence_score: number;
  summary: string | null;
  positive_signals: string[] | null;
  red_flags: string[] | null;
  data_gaps: string[] | null;
  scoring_breakdown: Record<string, { score: number; reasoning: string }> | null;
  draft_subject: string | null;
  draft_body: string | null;
  draft_type: "rejection" | "meeting_request" | null;
  research_sources: string[] | null;
  user_override: string | null;
  user_override_at: string | null;
  user_rating: "up" | "down" | null;
  user_rating_at: string | null;
  approved_at: string | null;
  sent_at: string | null;
  created_at: string;
}

export interface PaginatedLeads {
  total: number;
  page: number;
  page_size: number;
  items: Lead[];
}
